server/pairing/challenge_management.go

```diff
@@ -0,0 +1,247 @@
+	}
+	cg.authedIP = hIP
+	return nil
+}
+
+func (cg *ChallengeGiver) validateClientIP(r *http.Request) *ChallengeError {
+	// If we haven't registered yet register the IP
+	if cg.authedIP == nil || len(cg.authedIP) == 0 {
+		err := cg.registerClientIP(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Then compare the current req RemoteIP with the authed IP
+	hIP, err := cg.getIP(r)
+	if err != nil {
+		return err
+	}
+
+	if !cg.authedIP.Equal(hIP) {
+		cg.logger.Error(
+			"request RemoteAddr does not match authedIP: expected '%s', received '%s'",
+			zap.String("expected", cg.authedIP.String()),
+			zap.String("received", hIP.String()),
+		)
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+	return nil
+}
+
+func (cg *ChallengeGiver) getSession(r *http.Request) (*sessions.Session, *ChallengeError) {
+	s, err := cg.cookieStore.Get(r, sessionChallenge)
+	if err != nil {
+		cg.logger.Error("checkChallengeResponse: cg.cookieStore.Get(r, sessionChallenge)", zap.Error(err), zap.String("sessionChallenge", sessionChallenge))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return s, nil
+}
+
+func (cg *ChallengeGiver) generateNewChallenge(s *sessions.Session, w http.ResponseWriter, r *http.Request) ([]byte, *ChallengeError) {
+	challenge := make([]byte, 64)
+	_, err := rand.Read(challenge)
+	if err != nil {
+		cg.logger.Error("regenerateNewChallenge: _, err = rand.Read(challenge)", zap.Error(err))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	s.Values[sessionChallenge] = challenge
+	err = s.Save(r, w)
+	if err != nil {
+		cg.logger.Error("regenerateNewChallenge: err = s.Save(r, w)", zap.Error(err))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return challenge, nil
+}
+
+func (cg *ChallengeGiver) block(s *sessions.Session, w http.ResponseWriter, r *http.Request) *ChallengeError {
+	s.Values[sessionBlocked] = true
+	err := s.Save(r, w)
+	if err != nil {
+		cg.logger.Error("block: err = s.Save(r, w)", zap.Error(err))
+		return &ChallengeError{"error", http.StatusInternalServerError}
+	}
+
+	return &ChallengeError{"forbidden", http.StatusForbidden}
+}
+
+func (cg *ChallengeGiver) checkChallengeResponse(w http.ResponseWriter, r *http.Request) *ChallengeError {
+	ce := cg.validateClientIP(r)
+	if ce != nil {
+		return ce
+	}
+
+	s, ce := cg.getSession(r)
+	if ce != nil {
+		return ce
+	}
+
+	blocked, ok := s.Values[sessionBlocked].(bool)
+	if ok && blocked {
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+
+	// If the request header doesn't include a challenge don't punish the client, just throw a 403
+	pc := r.Header.Get(sessionChallenge)
+	if pc == "" {
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+
+	c, err := cg.encryptor.decryptPlain(base58.Decode(pc))
+	if err != nil {
+		cg.logger.Error("checkChallengeResponse: cg.encryptor.decryptPlain(base58.Decode(pc))", zap.Error(err), zap.String("pc", pc))
```

Comment:
@Samyoul, It almost feels like we're trying to build a stack trace of sorts because the error messages are basically tracking the source code 1:1. Doesn't `err` contain enough information that we can grasp what happened from the logs? By default `zap` annotate all logs with the functions' file name and line number. If the logs do contain such annotations, at least to me it seems like we can reduce the verbosity of the log messages, because maintaining these logs also have a cost, especially since they're very detailed (copy from source).

Do we have any kind of guideline on `status-go` about how we prefer to log errors and the rationale?

Reply:
This is a bit of my own creation. `zap` logs *can* show the full stacktrace along with a lot of other very helpful metadata, however the logger isn't always configured to show this helpful information. We have a lot of different logger instantiations and most put the logger in the least verbose mode. Copying the source into the log message is my way to try to even this inconsistency.

Ultimately I could put a descriptive message in the text field, but I would only be using it to describe in every day speech what the source code does, which is human readable. Also I'd try to make the comment as unique as possible so that I could grep it easily.

Better logging is something that `status-go` needs, we should be taking full advantage of `zap`'s functionality so I propose that:

- I change logger messages in this PR to something human readable
- I create an issue to plan and track upgrading logging generally across all `status-go` to use more featureful logging.
- We along with anyone else that cares formalises logging guidelines

Does this seem reasonable to you @ilmotta?

Reply:
Sounds like a most excellent plan @Samyoul.

Given the way we log will soon/eventually improve based on the plan you outlined above, I'm totally fine if you keep the logs the way they are in this PR and any other ones coming after :) Until the new logging guidelines are brought to life it might be better to stick to what we've been doing, even if there are multiple conventions.

I was just not on the same page as you in terms of the current logging situation in `status-go`, but now I understand better your rationale for logging the way this PR does.

Thank you!